metricflow/model/semantics/linkable_spec_resolver.py

```diff
@@ -533,10 +578,83 @@ def _get_semantic_models_with_joinable_entity(
                 valid_semantic_models.append(semantic_model)
         return valid_semantic_models
 
+    @staticmethod
+    def _get_time_granularity_for_dimension(
+        semantic_model: SemanticModel, time_dimension_reference: TimeDimensionReference
+    ) -> TimeGranularity:
+        matching_dimensions = tuple(
+            dimension
+            for dimension in semantic_model.dimensions
+            if dimension.type is DimensionType.TIME and dimension.time_dimension_reference == time_dimension_reference
+        )
+
+        if len(matching_dimensions) != 1:
+            raise RuntimeError(
+                f"Did not find a matching time dimension for {time_dimension_reference} in {semantic_model}"
+            )
+        time_dimension = matching_dimensions[0]
+        type_params = time_dimension.type_params
+        assert (
+            type_params is not None
+        ), f"type_params should have been set for {time_dimension_reference} in {semantic_model}"
+
+        assert (
+            type_params.time_granularity is not None
+        ), f"time_granularity should have been set for {time_dimension_reference} in {semantic_model}"
+
+        return type_params.time_granularity
+
     def _get_linkable_element_set_for_measure(self, measure_reference: MeasureReference) -> LinkableElementSet:
         """Get the valid linkable elements for the given measure."""
         measure_semantic_model = self._get_semantic_model_for_measure(measure_reference)
 
+        # Create elements related to the metric_time
+        measure_agg_time_dimension_reference = measure_semantic_model.checked_agg_time_dimension_for_measure(
```

Comment:
All of this new logic feels like it should be a separate helper method with a good name and a nice docstring. None of it requires anything local and it adds a bunch of complexity here which makes the core method harder to read.

Reply:
Separated this out into a separate method, but also moved out some other sections as the function was pretty long.